tools/difftags/src/page.rs

```diff
@@ -0,0 +1,234 @@
+            File::Modified {
+                old_name: source.into(),
+                new_name: target.into(),
+                sections,
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Section {
+    pub start_line: (usize, usize),
+    pub context_prefix: Option<Vec<Line>>,
+    pub diff: Vec<Line>,
+    pub context_suffix: Option<Vec<Line>>,
+    pub end_line: (usize, usize),
+}
+
+impl From<&Hunk> for Section {
+    fn from(hunk: &Hunk) -> Self {
+        assert!(!hunk.lines().is_empty());
+        let mut diff_start = None;
+        let mut suffix_start = None;
+        for (index, line) in hunk.lines().iter().enumerate() {
+            if diff_start.is_none() {
+                if line.line_type != " " {
+                    diff_start = Some(index);
+                }
+            } else if suffix_start.is_none() && line.line_type == " " {
+                suffix_start = Some(index);
+            }
+        }
+
+        let diff_start = diff_start.unwrap().saturating_sub(10);
+        let suffix_start = usize::min(
+            hunk.lines().len(),
+            suffix_start
+                .unwrap_or_else(|| hunk.lines().len())
+                .saturating_add(10),
+        );
+
+        let context_prefix: Vec<Line> = (&hunk.lines()[0..diff_start]).into();
+        let lines: Vec<Line> = (&hunk.lines()[diff_start..suffix_start]).into();
+        let context_suffix: Vec<Line> = (&hunk.lines()[suffix_start..]).into();
+        let end_line = &hunk.lines()[hunk.lines().len() - 1];
+
+        Self {
+            start_line: (
+                hunk.lines()[0].source_line_no.unwrap_or_default(),
+                hunk.lines()[0].target_line_no.unwrap_or_default(),
+            ),
+            context_prefix: if context_prefix.is_empty() {
+                None
+            } else {
+                Some(context_prefix)
+            },
+            diff: lines,
+            context_suffix: if context_suffix.is_empty() {
+                None
+            } else {
+                Some(context_suffix)
+            },
+            end_line: (
+                end_line.source_line_no.unwrap_or_default(),
+                end_line.target_line_no.unwrap_or_default(),
+            ),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use unidiff::PatchSet;
+
+    #[test]
+    fn test_hunk_to_section() {
+        let diff_str = r#"
+diff --git a/some/path/to/file.rs b/some/path/to/file.rs
+index 422b64415..9561909ed 100644
+--- a/some/path/to/file.rs
++++ b/some/path/to/file.rs
+@@ -1,31 +1,31 @@
+ 00
+ 01
+ 02
+ 03
+ 04
+ 05
+ 06
+ 07
+ 08
+ 09
+ 10
+ 11
+ 12
+ 13
+ 14
+-oops
++15
+ 16
+ 17
+ 18
+ 19
+ 20
+ 21
+ 22
+ 23
+ 24
+ 25
+ 26
+ 27
+ 28
+ 29
+ 30
+        "#;
+        let mut patch = PatchSet::new();
+        patch.parse(&diff_str).unwrap();
+
+        let hunk = &patch.files()[0].hunks()[0];
+        let section: Section = hunk.into();
+
+        assert_eq!((1, 1), section.start_line);
```

Comment:
Can we explain in the comment how the the fields in `Section` are derived from the test data above? I can sort of guess from the code and `assert_eq!` but am not 100% sure, e.g., how the `section.diff` starts at `05` and ends at `25`.

Reply:
Yeah, good call. I created a constant for the number of lines surrounding the modified lines that will be displayed, commented it, and also added additional context to the assert messages.